### ISSUE TITLE
Going full algebraic type with Enum's

### ISSUE THREAD

**Issue by dmoisset** [gvanrossum/patma#116]
Sorry for the length of this, but this is an idea that requires some build-up, otherwise you'll likely hate it at first sight. If you think this is crazy in your first read, read it twice. :) Some of the argumentation of this come from my [notes](https://github.com/dmoisset/notebook/blob/master/python/pep622/understanding-pep-622.md), but the short version is that I'm reapproaching some of the pattern syntax based on the core idea that we're implementing algebraic types in python.

Algebraic types are a CS concept, not a Python concept, so PEP-622 is trying to map between those; this issue proposes an alternate mapping to the one implicitly in the PEP which has some different benefits and drawbacks. Some of that addresses big parts of #90, but I wanted to put this separately as it covers a wider issue.

A CS algebraic type is built on the disjoint union of the results of some "constructors", which are functions more or less isomorphic to tuple construction plus a label to ensure the union is actually disjoint. A normal interpretation of dynamic type systems is that each runtime value has the label of the class that built it (the `__class__` slot in python) making it work like a single big algebraic data type. The existing class_pattern fits that idea perfectly, and uses class names as labels/constructors and matches with that then doing some transformation on the content via `__match_args__` to abstractly turn into a sort of namedtuple. I'll ignore that second part (deconstruction) in this issue which only focuses on constructors. The sealed class idea goes further that way (an abstract parent class is the algebraic type, and each concrete subclass one of the variants). The classes themselves in Python are constructor functions which makes a lot of sense in this conceptual mapping.

In most languages with algebraic data types, enumerations are modeled in the same way: In Haskell if you define a data type `data Size = Small | Medium | Big`, the 3 sizes are semantically functions with 0 arguments and are called "constructors". So they use the same language features (and pattern matching) as other more complex types. Python `Enum`s are somewhat different, and its elements are not constructors. In a parallel universe if I have an `Size = enum.Enum("Size", "Small Medium Big")`, I could end up with a `Size` class with a `Small()` classmethod / constructor that would return me an instance of `Size` (as do most other python classes with multiple constructors). As we live in this universe and want to be able to match 2 different implementations of algebraic types that coexist in python (the one for classes and the one for enums), PEP 622 ended up having two different constructor matches: the Class Pattern for objects-with-class-as-label and the (dotted in the middle) Constant Value Pattern for enums (and enum-likes). 

The proposal of this issue is to instead implement the `match` statement from that parallel universe, even if we look in this. Than means we would get:

```python
match hat.size:
    case Size.Small(): ...
    case Size.Medium(): ...
    case Size.Big(): ...
``` 

Yes, this breaks a bit the notion of Syntax Mirroring (as defined on #101), but all the alternatives we have on #90 break it a bit anyway. Or you could extend this and make `Enum` instances callable and self-returning if you want to keep this and also make Python more aligned with an unified algebraic type vision. (This paragraph is an optional and unrequired part of the proposal, and I actually prefer not to do it.)

At this point you might ask "what's the gain here, besides satisfying some academic notion of equivalence between two python concepts"? After doing this, we no longer need the "Constant Value Pattern" (the remaining cases can be relegated to `if/else`, guards, or be more bold as I describe below).

The "Class Pattern" still has exactly the same syntax (although it may be renamed as "Constructor Pattern"). The semantic evaluates the pseudo-expression on the left of the `(` ; if it is a `type`, the meaning is as it is now. if it's not, the parenthesized list must be empty (throwing TypeError otherwise) and the non-type value is compared with equality to the object being matched. I'm stretching a bit now and saying that a module attribute like `logging.INFO()` is also a constructor. And if we stretch it even more, then `MY_GLOBAL_CONSTANT()` can be seen as the constructor for the value in `MY_GLOBAL_CONSTANT`. (This paragraph is an optional and unrequired part of the proposal, but I would put this in on the first version), and this would cover all the cases we lost.

With this:
* the way to match constant is something that unambiguously looks unlike an assignment target, and I think it makes more obvious that you're not assigning to it
* The only load semantics left now are "whatever is on the left of the `(` in a class pattern is interpreted as a value". The only things that still look assignable, actually are. :)

If you saw my syntax table in my previous article, this is the new one:

|Pattern               |example       |target-like|expression-like|binds a variable|
|----------------------|--------------|:---------:|:-------------:|----------------|
| name_pattern         |`foo`         |✔️|✔️|✔️|
| literal_pattern      |`42`          |✖️|✔️|✖️|
| sequence_pattern     |`[x,y]`       |✔️*|✔️*|✔️†|
| mapping_pattern      |`{"x": x}`    |✖️|✔️*|✔️†|
| class_pattern        |`Point(x,y)`  |✖️|✔️*|✔️†|

I could even go further and propose making the mapping and class patterns valid destructuring targets usable in other statements (assignment, for, etc). With that the syntax table would be extremely homogeneous (the literal_pattern being the only one not ticking all the boxes, but also the one where no one has said to be confused about its meaning). And IMO that helps ensuring it's  simple understanding: iff a pattern looks like an assignment target, it works like it. (This paragraph is an optional and unrequired part of the proposal.)

As a drawback, that still doesn't cover the problem of someone forgetting the (), but if a pattern match looks much closer to an assignment with an extra "only assign if it fits", and we can leave the rest of the prevention of the shadowing problems to linters/type checkers.

**Comment by gvanrossum** [comment]
Intriguing. Even if we can implement this for `Enum`, this will leave more pedestrian approaches to naming constants in the dust, and getting enough people to upgrade their named constants to enums to make that a non-issue would take a very long time. So I expect this won't fly from a "practicality beats purity" issue.

Ideally I would want something that people can plug into their code as soon as they upgrade to 3.10, without first having to refactor their whole code base (let alone dependencies).

**Comment by dmoisset** [comment]
I don't think any refactor would be needed. My "practicality beats purity" approach was giving the semantics I explained to the Constructor Pattern that makes this work

```
match loglevel:
     case logging.INFO(): ...
```

WITHOUT requiring changes to the logging module (otherwise this proposal would be a massive break to backwards compatibility).

**Comment by gvanrossum** [comment]
Ah, sorry. Then the main concern is whether we can actually get people to write `case logging.INFO()`. The presence of literal patterns will forever confuse people, because `if target == 0` translates into `case 0` but `if target == INFO` would have to become `case INFO()`.

**Comment by viridia** [comment]
One of the advantages of the current "if it contains a dot, it's a value pattern rather than a capture pattern" heuristic is that it "just works" for the vast majority of cases. My example program, `expr.py`, contained half a dozen `match` statements, each having somewhere between 5-10 case clauses, many with subpatterns - and not one of them used an unqualified constant name requiring a leading dot (e.g. .BLACK). So at least in the code **I** write, such constructions are extremely rare.

Now it is true that this observation hinges on the fact that I imported the `tokenize` module as `import tokenize` rather than `from tokenize import *`. This in turn means that the few places where I use named constants in `match`, I am writing cases of the form `case [tokenize.OP, value]:` instead of `case [OP, value]:`. I can imagine that some programmers might choose differently, which means that the use of unqualified constants might be more common in other people's code than in my own. But to me, using enum members in this way feels like a code smell - it increases the chance that one of the imported names will collide with some other name without my knowing it.

As an aside, I want to mention that in TypeScript, the use of unions of singleton types is favored over string-valued enums in most cases. Thus you often see code like this:

```
type Size = 'small' | 'medium' | 'large';

function Button(size: Size) { /* ... */ }

const Button = new Button('small');
```

In other words, we're writing a literal instead of a symbol, which is more succinct than something like `Size.SMALL`, but because the parameter is statically typed the literal has all the same type-checking and editor auto-completion benefits that using an enum would have, and is just as readable if not more so. However, none of these arguments apply to *number*-valued enums, which are still encouraged to be enums.

Anyway, my point is this: regardless of what we decide to do with unqualified names, I think we should retain the "if it contains a dot and is not followed by parens, then it's a constant value" rule for qualified names - simply because it's both intuitive for the beginner and clean to look at.

**Comment by viridia** [comment]
I like the name "constructor pattern" BTW.

**Comment by dmoisset** [comment]
> Ah, sorry. Then the main concern is whether we can actually get people to write `case logging.INFO()`. The presence of literal patterns will forever confuse people, because `if target == 0` translates into `case 0` but `if target == INFO` would have to become `case INFO()`.

Yes, now we agree that's the main concern. I know the proposal is... different. I wanted to throw it here to see if it sticks, and if that happens we might discuss about how to convince people. If people in this repo are not buying it then we cose this, and thanks for reading.

**Comment by gvanrossum** [comment]
Yeah, let’s not go here.
